File: java/src/main/java/ai/smritea/sdk/model/AddOptions.java
package ai.smritea.sdk.model;

import java.util.Map;

/**
 * Builder-style options for adding a memory.
 *
 * <p>Usage example:
 *
 * <pre>{@code
 * new AddOptions()
 *     .withScope(MemoryScope.builder().actorId("alice").actorType("user").build())
 *     .withMetadata(Map.of("key", "value"));
 * }</pre>
 */
public final class AddOptions {
  private MemoryScope scope;
  private Map<String, Object> metadata;
  private String eventOccurredAt;
  private RelativeStanding relativeStanding;

  /** Creates a new AddOptions with all fields unset. */
  public AddOptions() {}

  /**
   * Sets the memory scope (actor and conversation context).
   *
   * @param scope the memory scope
   * @return this instance for chaining
   */
  public AddOptions withScope(MemoryScope scope) {
    this.scope = scope;
    return this;
  }

  /**
   * Sets arbitrary key-value metadata.
   *
   * @param metadata the metadata map
   * @return this instance for chaining
   */
  public AddOptions withMetadata(Map<String, Object> metadata) {
    this.metadata = metadata;
    return this;
  }

  /**
   * Sets the temporal anchor for extraction (ISO-8601 datetime string). Used by the extraction LLM
   * to resolve relative temporal expressions like "last year" or "yesterday". Defaults to current
   * time if omitted.
   *
   * @param eventOccurredAt ISO-8601 datetime string
   * @return this instance for chaining
   */
  public AddOptions withEventOccurredAt(String eventOccurredAt) {
    this.eventOccurredAt = eventOccurredAt;
    return this;
  }

  /**
   * Sets importance and temporal decay configuration for this memory.
   *
   * @param relativeStanding the relative standing configuration
   * @return this instance for chaining
   */
  public AddOptions withRelativeStanding(RelativeStanding relativeStanding) {
    this.relativeStanding = relativeStanding;
    return this;
  }

  /** Returns the memory scope. */
  public MemoryScope getScope() {
    return scope;
  }

  /** Returns the metadata map. */
  public Map<String, Object> getMetadata() {
    return metadata;
  }

  /** Returns the event occurred at timestamp. */
  public String getEventOccurredAt() {
    return eventOccurredAt;
  }

  /** Returns the relative standing configuration. */
  public RelativeStanding getRelativeStanding() {
    return relativeStanding;
  }
}

File: csharp/AddOptions.cs
// <copyright file="AddOptions.cs" company="PlaceholderCompany">
// Copyright (c) PlaceholderCompany. All rights reserved.
// </copyright>

namespace Smritea.Sdk;

/// <summary>
/// Builder-style options for the AddAsync method.
/// Use fluent <c>With*</c> methods for ergonomic construction:
/// <code>new AddOptions().WithScope(new MemoryScope().WithActorId("alice").WithActorType("user"))</code>
/// </summary>
public sealed class AddOptions
{
    /// <summary>Gets the scope containing actor and conversation context.</summary>
    public MemoryScope? Scope { get; private set; }

    /// <summary>Gets the arbitrary key-value metadata.</summary>
    public Dictionary<string, object>? Metadata { get; private set; }

    /// <summary>Gets the ISO-8601 datetime string for when this content was created or occurred.</summary>
    public string? EventOccurredAt { get; private set; }

    /// <summary>Gets the importance and temporal decay configuration.</summary>
    public RelativeStanding? RelativeStanding { get; private set; }

    /// <summary>Sets the scope containing actor and conversation context.</summary>
    /// <param name="scope">The scope object grouping actor and conversation fields.</param>
    /// <returns>The current instance for method chaining.</returns>
    public AddOptions WithScope(MemoryScope scope)
    {
        this.Scope = scope;
        return this;
    }

    /// <summary>Sets the metadata dictionary attached to the memory.</summary>
    /// <param name="metadata">Arbitrary key-value pairs to store alongside the memory.</param>
    /// <returns>The current instance for method chaining.</returns>
    public AddOptions WithMetadata(Dictionary<string, object> metadata)
    {
        this.Metadata = metadata;
        return this;
    }

    /// <summary>Sets the temporal anchor for extraction (ISO-8601 datetime string).
    /// Used by the extraction LLM to resolve relative temporal expressions like "last year".
    /// Defaults to current time if omitted.</summary>
    /// <param name="eventOccurredAt">ISO-8601 datetime string.</param>
    /// <returns>The current instance for method chaining.</returns>
    public AddOptions WithEventOccurredAt(string eventOccurredAt)
    {
        this.EventOccurredAt = eventOccurredAt;
        return this;
    }

    /// <summary>Sets importance and temporal decay configuration for this memory.</summary>
    /// <param name="relativeStanding">The relative standing configuration.</param>
    /// <returns>The current instance for method chaining.</returns>
    public AddOptions WithRelativeStanding(RelativeStanding relativeStanding)
    {
        this.RelativeStanding = relativeStanding;
        return this;
    }
}
